### PR TITLE
Allow clients to specify arbitrary overrides of operator-managed resources

### PR DESCRIPTION
**Pull request by ecordell**
Discussed in #61, this adds a `patches` field to the `SpiceDBCluster` object that allows arbitrary patching of operator-generated resources.

Currently, every object the operator creates per-SpiceDBCluster can be uniquely identified by its `kind`. So for now, `kind` is the only selector used to apply patches. We can add more filter options in the future if we ever need to resolve ambiguity (i.e. a label selector).

Patches can be defined as structured merge patches, where you just specify the fields you want to change, or as json6902 patch operations. 

Multiple patches can be specified for the same object, with later patches applying over top of earlier patches.

## Examples

Structured merge patch:
```yaml
apiVersion: authzed.com/v1alpha1
kind: SpiceDBCluster
metadata:
  name: dev
spec:
  config:
    datastoreEngine: memory
  secretName: dev-spicedb-config
  patches:
  - kind: Deployment
    patch:
      metadata:
        labels:
          added: via-patch 
      spec:
         template:
           metadata:
             labels: 
               added: pod-label-via-patch
```

Explicit JSON6902 Patch:
```yaml
apiVersion: authzed.com/v1alpha1
kind: SpiceDBCluster
metadata:
  name: dev
spec:
  config:
    datastoreEngine: memory
  secretName: dev-spicedb-config
  patches:
  - kind: Deployment
    patch:
      op: add
      path: /metadata/labels
      value: 
        added: via-patch
```

You can specify multiple patches for the same object (later in the list are applied over top of earlier in the list):

```yaml
apiVersion: authzed.com/v1alpha1
kind: SpiceDBCluster
metadata:
  name: dev
spec:
  config:
    datastoreEngine: memory
  secretName: dev-spicedb-config
  patches:
  - kind: Deployment
    patch:
      op: add
      path: /metadata/labels
      value:
        added: via-patch
  - kind: Deployment
    patch:
      metadata:
        labels:
          added-2: via-patch-2
```

Apply a label to all objects the operator creates:

```yaml
apiVersion: authzed.com/v1alpha1
kind: SpiceDBCluster
metadata:
  name: dev
spec:
  config:
    datastoreEngine: memory
  secretName: dev-spicedb-config
  patches:
  - kind: '*'
    patch:
      op: add
      path: /metadata/labels
      value:
        added: via-wildcard-patch
```

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -383,31 +421,37 @@ func (c *Config) ToEnvVarApplyConfiguration() []*applycorev1.EnvVarApplyConfigur
...
 		WithLabels(metadata.LabelsForComponent(c.Name, metadata.ComponentServiceAccountLabel)).
 		WithAnnotations(c.ExtraServiceAccountAnnotations).
-		WithOwnerReferences(c.OwnerRef())
+		WithOwnerReferences(c.ownerRef())
 }
 
-func (c *Config) Role() *applyrbacv1.RoleApplyConfiguration {
+func (c *Config) ServiceAccount() *applycorev1.ServiceAccountApplyConfiguration {
+	sa := c.serviceAccount()
+	_ = ApplyPatches("ServiceAccount", sa, c.Patches)
```
Comment (vroldanbet) [inline pkg/config/config.go:447]: it would seem like we are applying the patch 2 times? one in `NewConfig` and one in the individual resource methods?

File: pkg/config/patch.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	for _, p := range patches {
+		if p.Target.Kind == kind {
+			// marshal object to json
+			encoded, err := json.Marshal(object)
+			if err != nil {
+				return err
+			}
+
+			// determine if the patch is a strategic merge or a json6902 patch
+			decoder := utilyaml.NewYAMLOrJSONDecoder(strings.NewReader(p.Patch), 100)
```
Comment (vroldanbet) [inline pkg/config/patch.go:27]: what is that 100 number?

Comment (ecordell) [reply pkg/config/patch.go:27]: the number of bytes to look ahead to try and determine whether it's json or yaml

File: pkg/config/config.go (hunk 1 above)
Comment (ecordell) [reply pkg/config/config.go:447]: in `NewConfig` it's just for validation so that we can show good errors, the actual application happens later. if we decide to go this route, I will clean it up so that it only processes once.

File: pkg/config/patch_test.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			if _, _, err := ApplyPatches(tt.object, tt.patches); (err != nil) != tt.wantErr {
+				t.Errorf("ApplyPatches() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			require.Equal(t, tt.want, tt.object)
+		})
+	}
+}
+
+// TODO: test that all exported methods on `Config` apply patches
```
Comment (vroldanbet) [inline pkg/config/patch_test.go:148]: just in case you intended to address this in the PR ☝🏻

File: pkg/config/patch_test.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+			}},
+			want: applyappsv1.Deployment("test", "test").WithLabels(
+				map[string]string{
+					"initial": "label",
+					"added":   "via-patch",
+				},
+			),
+		},
+		{
+			name:   "add labels to deployment with a wildcard match (smp, yaml)",
```
Comment (vroldanbet) [inline pkg/config/patch_test.go:125]: can we add a patch that does not specify its `kind`?

File: pkg/config/patch_test.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+metadata:
+  labels:
+    added: via-patch`),
+			}},
+			want: applyappsv1.Deployment("test", "test").WithLabels(map[string]string{"added": "via-patch"}),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if _, _, err := ApplyPatches(tt.object, tt.patches); (err != nil) != tt.wantErr {
```
Comment (vroldanbet) [inline pkg/config/patch_test.go:140]: shouldn't we also test the other arguments to make sure `ApplyPatches` reports the correct number of applied patches and wether a diff existed? A potential test would be to apply 2 patches, one of which fails, so we see `diff = true` and `count = 1`

**Review by vroldanbet — APPROVED** [review]
This looks good to me, I left a handful of questions, but I don't consider any of them blockers. I took the liberty to set the base branch to `graph-test-precalc` to reduce the size of the diff and facilitate the review 🙏🏻

File: pkg/controller/validate_config.go [hunk 6]
```diff
@@ -71,15 +58,15 @@ func (c *ValidateConfigHandler) Handle(ctx context.Context) {
...
-		CurrentMigrationHash: currentStatus.Status.CurrentMigrationHash,
-		SecretHash:           currentStatus.Status.SecretHash,
+		CurrentMigrationHash: cluster.Status.CurrentMigrationHash,
+		SecretHash:           cluster.Status.SecretHash,
 		Image:                validatedConfig.TargetSpiceDBImage,
 		Migration:            validatedConfig.TargetMigration,
 		Phase:                validatedConfig.TargetPhase,
 		CurrentVersion:       validatedConfig.SpiceDBVersion,
-		Conditions:           *currentStatus.GetStatusConditions(),
+		Conditions:           *cluster.GetStatusConditions(),
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/controller/validate_config.go:69]: It's used by the `Pause` handler, which is in `controller-idioms` and generic over the API type, which may be why your tooling can't find the usages?

File: pkg/config/config.go [hunk 7]
```diff
@@ -351,20 +353,51 @@ func NewConfig(cluster *v1alpha1.SpiceDBCluster, globalConfig *OperatorConfig, s
...
+	out := &Config{
+		MigrationConfig: migrationConfig,
+		SpiceConfig:     spiceConfig,
+		Patches:         cluster.Spec.Patches,
+	}
+
+	// Validate that patches apply cleanly ahead of time
+	totalAppliedPatches := 0
+	for _, obj := range []any{
+		out.serviceAccount(),
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/config/config.go:365]: There's a test that confirms all exported methods return an object, and that patches are applied to all objects generated by the config. I think that's a stronger guarantee than turning this into a method?

File: pkg/config/config.go [hunk 8]
```diff
@@ -391,31 +424,37 @@ func (c *Config) ToEnvVarApplyConfiguration() []*applycorev1.EnvVarApplyConfigur
...
 		WithLabels(metadata.LabelsForComponent(c.Name, metadata.ComponentServiceAccountLabel)).
 		WithAnnotations(c.ExtraServiceAccountAnnotations).
-		WithOwnerReferences(c.OwnerRef())
+		WithOwnerReferences(c.ownerRef())
 }
 
-func (c *Config) Role() *applyrbacv1.RoleApplyConfiguration {
+func (c *Config) ServiceAccount() *applycorev1.ServiceAccountApplyConfiguration {
+	sa := c.serviceAccount()
+	_, _, _ = ApplyPatches(sa, c.Patches)
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/config/config.go:450]: Yeah, I considered ways to only call the patching once, but couldn't find anything that I thought was substantially better.

I like the naming suggestions, I will update.

File: pkg/controller/client.go [hunk 9]
```diff
@@ -14,6 +14,7 @@ func (c *Controller) PatchStatus(ctx context.Context, patch *v1alpha1.SpiceDBClu
 	for _, c := range patch.Status.Conditions {
 		c.ObservedGeneration = patch.Generation
 	}
+	patch.ManagedFields = nil
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/controller/client.go:17]: I meant to put this into a separate commit for clarity (and I will update to do that)

The mechanical reason is just that `apply` fails if you have managed fields set. The apiserver uses the value send as the "field manager" to set the managed field values for you, and doesn't let you directly overwrite them.

But the reason this change is in this PR at all is because I combined the `CtxCluster` object and the `CtxClusterStatus` object into one. 

The original idea was that `CtxCluster` was a reference to the `SpiceDBCluster` in the informer cache (which shouldn't be modified) so any handler could recover the "original" object being synced if it needed it. And the `CtxClusterStatus` was a copy of the `SpiceDBCluster` object that would be continually updated if we made changes to it during a sync, so that subsequent handlers would get the "new" view of the object.

But in practice I thought it was hard to remember the distinction, and I just changed `CtxCluster` to be a deepcopy of the object from the cache, and it's easy enough to query the cache directly if we need to recover the original object in the future.

File: pkg/controller/validate_config.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (vroldanbet) [reply pkg/controller/validate_config.go:69]: let me rephrase it. I couldn't wrap my head around needing to return a pointer to a slice, so I tried to find usages that could justify it. I did find usages, but none seemed to explain the choice for pointer to slice. Mind clarifying?

File: pkg/config/patch.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+	count := 0
+	errs := make([]error, 0)
+	for i, p := range patches {
+		if p.Kind == *typeMeta.Kind || p.Kind == wildcard {
+			// determine if the patch is a strategic merge or a json6902 patch
+			decoder := utilyaml.NewYAMLOrJSONDecoder(bytes.NewReader(p.Patch), 100)
+			var json6902op jsonpatch.Operation
+			err = decoder.Decode(&json6902op)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("error decoding patch %d: %w", i, err))
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/config/patch.go:57]: json decoding doesn't fail if the expected fields don't match, just if there's a structural issue with the json (or yaml) itself

File: pkg/config/patch_test.go (hunk 4 above)
Comment (ecordell) [reply pkg/config/patch_test.go]: Added a lot more tests here

File: pkg/config/patch_test.go (hunk 3 above)
Comment (ecordell) [reply pkg/config/patch_test.go]: Yep, it's in `config_test.go`

File: pkg/config/config.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (vroldanbet) [reply pkg/config/config.go]: it seems good enough for me!

**Review by vroldanbet — APPROVED** [review]

**Comment by ecordell** [comment]
(I switched the base branch back to main, the commit content is the same)

**Review by vroldanbet — APPROVED** [review]

**Review by vroldanbet — APPROVED** [review]
